src/algorithm/bluestein.rs

```diff
@@ -109,45 +104,42 @@ fn calculate_x_twiddles<T: FFTnum>(
+    /// The inner FFT determines if this becomes a forward or inverse transform.
     ///
     /// Note that this constructor is quite expensive to run; This algorithm must run a FFT within the
     /// constructor. 
-    pub fn new(len: usize, inner_fft_fw: Arc<FFT<T>>, inner_fft_inv: Arc<FFT<T>>, inverse: bool) -> Self {
-        let inner_fft_len = (2 * len - 1).checked_next_power_of_two().unwrap();
-        assert_eq!(inner_fft_len, inner_fft_fw.len(), "For Bluesteins algorithm, inner_fft_fw.len() must be a power of to larger than or equal to 2*self.len() - 1. Expected {}, got {}", inner_fft_len, inner_fft_fw.len());
-        assert_eq!(inner_fft_len, inner_fft_inv.len(), "For Bluesteins algorithm, inner_fft_inv.len() must be a power of to larger than or equal to 2*self.len() - 1. Expected {}, got {}", inner_fft_len, inner_fft_inv.len());
+    pub fn new(len: usize, inner_fft: Arc<FFT<T>> ) -> Self {
+        let min_inner_fft_len = 2 * len - 1;
+        assert!(inner_fft.len() >= min_inner_fft_len, "For Bluesteins algorithm, inner_fft.len() must be equal to or larger than 2*self.len() - 1. Expected at least {}, got {}", min_inner_fft_len, inner_fft.len());
 
-        let mut w_twiddles = vec![Complex::zero(); inner_fft_len];
+        let mut w_twiddles = vec![Complex::zero(); inner_fft.len()];
         let mut x_twiddles = vec![Complex::zero(); len];
-        calculate_w_twiddles(len, &inner_fft_fw, &mut w_twiddles, inverse);
-        calculate_x_twiddles(len, &mut x_twiddles, inverse);
+        calculate_w_twiddles(len, &inner_fft, &mut w_twiddles);
+        calculate_x_twiddles(len, &mut x_twiddles, inner_fft.is_inverse());
         Self {
             len,
-            inner_fft_fw,
-            inner_fft_inv,
+            inner_fft,
             w_twiddles: w_twiddles.into_boxed_slice(),
             x_twiddles: x_twiddles.into_boxed_slice(),
-            inverse,
         }
     }
 
     fn perform_fft(&self, input: &mut [Complex<T>], output: &mut [Complex<T>]) {
         assert_eq!(self.len(), input.len());
 
-        let mut scratch_a = vec![Complex::zero(); self.inner_fft_fw.len()];
-        let mut scratch_b = vec![Complex::zero(); self.inner_fft_fw.len()];
+        let mut scratch = vec![Complex::zero(); 2*self.inner_fft.len()];
+        let (mut scratch_a, mut scratch_b) = scratch.split_at_mut(self.inner_fft.len());
         for (w, (x, i)) in scratch_a.iter_mut().zip(self.x_twiddles.iter().zip(input.iter())) {
             *w = x * i;
         }
-        self.inner_fft_fw.process(&mut scratch_a, &mut scratch_b);
+        self.inner_fft.process(&mut scratch_a, &mut scratch_b);
         for (w, wi) in scratch_b.iter_mut().zip(self.w_twiddles.iter()) {
-            *w = *w * wi;
+            *w = (*w * wi).conj();
         }
-        self.inner_fft_inv.process(&mut scratch_b, &mut scratch_a);
+        self.inner_fft.process(&mut scratch_b, &mut scratch_a);
```

Comment:
I have one final request: Could you leave a comment here pointing out that this is computing an inverse FFT, because we have conjugated the inputs and outputs?

Reply:
I updated this now and added comments on what is going on.
I also added conjugations so the first FFT is always a forward type, and the second always an inverse. It seems using an iFFT as the first step and an FFT as the second also produces correct results. I haven't investigated why that is, but that's not very intuitive so I prefer to do it the "normal" way.